### PR TITLE
Refactor and Optimize FA3 Code

### PR DESCRIPTION
**Pull request by hebiao064**
## Motivation
- Accelerate Spec Decoding + FA3 + CUDA Graph
- Easier to maintain

Follow up on https://github.com/sgl-project/sglang/pull/5050

## Benchmark for Llama 8B (Before and After)

| Iterations | GSM Latency on H200 | Send One on H200 |
|-------------------|-----------------------|-----------------------|
| Main Branch with Flash Infer           |  2118.780 token/s                    |   229.67 token/s               | 
| Main Branch  with FA3          |  2000.892 token/s                    |  235.28 token/s                 | 
| FA3 After commit: [21bf3fe](https://github.com/sgl-project/sglang/pull/5090/commits/21bf3fed77c73185f0161a247bc5d76d4e65c830)  and [eb77c88](https://github.com/sgl-project/sglang/pull/5090/commits/eb77c884a416c4bbd2b463497f0a39d295e72bf2)   | 2428.587 token/s               |  235.00 token/s |

GSM Accuracy is Accuracy: 0.775 with spec decoding.

## Sanity Check without Spec Decoding for Llama 8B
Accuracy: 0.775
Invalid: 0.000
Latency: 4.473 s

Accuracy: 0.775
Invalid: 0.000
Latency: 3.993 s
Output throughput: 4255.736 token/s

## Sanity Check with Deepseek V3 (Send One)
- Deepseek V3 with flashinfer + cuda graph: speed=41.59 token/s
- Deepseek V3 with flashinfer + disabled cuda graph: speed=4.60 token/s
- Deepseek V3 with flashinfer + cuda graph + spec decode: acc_length=2.59, speed=62.73 token/s
- Deepseek V3 with flashinfer + disabled cuda graph + spec decode:  acc_length=2.59, speed=11.09 token/s

- Deepseek V3 with fa3 + cuda graph: speed=44.93 token/s, result is good
- Deepseek V3 with fa3 + disabled cuda graph: speed=4.22 token/s token/s, result is good
- Deepseek V3 with fa3 + cuda graph + spec decode: acc_length=2.64, speed=66.66 token/s
- Deepseek V3 with fa3 + disabled cuda graph + spec decode: acc_length=2.69, speed=11.72 token/s 

## Sanity Check with Deepseek V3 (GSM8K)
- Deepseek V3 with flashinfer + cuda graph:
Accuracy: 0.960
Invalid: 0.000
Latency: 45.092 s
Output throughput: 439.656 token/s

- Deepseek V3 with flashinfer + disabled cuda graph:
Accuracy: 0.970
Invalid: 0.000
Latency: 154.613 s
Output throughput: 126.697 token/s

- Deepseek V3 with flashinfer + cuda graph + spec decode: 
Accuracy: 0.970
Invalid: 0.000
Latency: 42.325 s
Output throughput: 461.995 token/s

- Deepseek V3 with flashinfer + disabled cuda graph + spec decode: 
Accuracy: 0.965
Invalid: 0.000
Latency: 93.745 s
Output throughput: 209.964 token/s

- Deepseek V3 with fa3 + cuda graph: 
Accuracy: 0.965
Invalid: 0.000
Latency: 43.330 s
Output throughput: 463.840 token/s

- Deepseek V3 with fa3 + disabled cuda graph: 
Accuracy: 0.950
Invalid: 0.000
Latency: 178.272 s
Output throughput: 112.082 token/s

- Deepseek V3 with fa3 + cuda graph + spec decode: 
Accuracy: 0.955
Invalid: 0.000
Latency: 36.043 s
Output throughput: 550.533 token/s

- Deepseek V3 with fa3 + disabled cuda graph + spec decode:
Accuracy: 0.960
Invalid: 0.000
Latency: 134.789 s
Output throughput: 148.343 token/s

## Modifications

See all commits msg, they are clear

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [x] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 1]
```diff
@@ -91,81 +87,67 @@ def __init__(
...
-                    self.step_id + 1
-                )
                 metadata.page_table = forward_batch.req_to_token_pool.req_to_token[
                     forward_batch.req_pool_indices, : metadata.max_seq_len_k
                 ]
-                cache_loc = forward_batch.out_cache_loc.view(
-                    self.speculative_num_steps, -1
-                ).T
-
-                for idx, single_seq_len in enumerate(seq_lens_with_decode):
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/flashattention_backend.py:343]: Can the for loop here be directly deleted?

Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:343]: yes, i verified it e2e and the metadata are the same before and after this for loop, it's only needed for top k > 1

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 2]
```diff
@@ -545,37 +509,30 @@ def init_forward_metadata_capture_cuda_graph(
                 )
             self.decode_cuda_graph_metadata[bs] = metadata
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/flashattention_backend.py:789]: Does this mean `self.decode_cuda_graph_metadata` has two different types of keys? (`string` for buffers and `int` for `bs`) It seems a little confusing.

Maybe the dict storing metadata can be seperated from the dict storing memory buffers. The same goes for `self.target_verify_graph_metadata`.

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 3]
```diff
@@ -442,7 +411,13 @@ def init_cuda_graph_state(self, max_bs: int):
         to avoid memory allocations.
         """
         self.decode_cuda_graph_metadata = {
-            # Page table for token mapping (batch_size, max_context_len)
+            "cache_seqlens": torch.zeros(max_bs, dtype=torch.int32, device=self.device),
+            "cu_seqlens_q": torch.arange(
+                0, max_bs + 1, dtype=torch.int32, device=self.device
+            ),
+            "cu_seqlens_k": torch.zeros(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/flashattention_backend.py:697]: Can `cu_seqlens_k` be removed from decode metadata? It's not used in the codes below.

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 4]
```diff
@@ -458,30 +433,22 @@ def init_cuda_graph_state(self, max_bs: int):
             "strided_indices": torch.arange(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/flashattention_backend.py:712]: Can `strided_indices` be removed from decode metadata? It's not used in the codes below.

File: python/sglang/srt/layers/attention/flashattention_backend.py [hunk 5]
```diff
@@ -458,30 +433,22 @@ def init_cuda_graph_state(self, max_bs: int):
...
                 0, self.max_context_len, self.page_size, device=self.device
             ),
+        }
+
+        self.target_verify_metadata = {
             "cache_seqlens": torch.zeros(max_bs, dtype=torch.int32, device=self.device),
-            "cu_seqlens_q": torch.arange(
-                0, max_bs + 128, dtype=torch.int32, device=self.device
+            "cu_seqlens_q": torch.zeros(
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/flashattention_backend.py:719]: Can `cu_seqlens_q` be removed from verify metadata? It's not used in the codes below.

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 4 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:712]: it is be needed for page size > 1 for decode

Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:712]: for target verify, it will be needed for page size > 1

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 2 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:789]: I think currently it's okay as we only need to metadata map, I'll consider about create two more metadata maps if we see maintainer keep get confused by this.

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 4 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:712]: https://github.com/sgl-project/sglang/pull/5168

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 5 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:719]: https://github.com/sgl-project/sglang/pull/5168

File: python/sglang/srt/layers/attention/flashattention_backend.py (hunk 3 above)
Comment (hebiao064) [reply python/sglang/srt/layers/attention/flashattention_backend.py:697]: https://github.com/sgl-project/sglang/pull/5168
